trunk/tkrev/tkrev.1
.TH tkrev 1
.SH NAME
\fBtkrev\fP - a Tk/Tcl Graphical Interface to CVS, Subversion and Git
.SH SYNOPSIS
.PP
\fBtkrev\fP [\fB-dir\fP \fIdirectory\fP] [\fB-root\fP \fIcvsroot\fP] [\fB-win workdir\fP|\fBmodule\fP|\fBmerge\fP]

\fBtkrev\fP [\fB-log\fP|\fBblame\fP \fIfile\fP]

\fBtkrev\fP \fBfile\fP - same as \fBtkrev -log\fP \fIfile\fP
.SH DESCRIPTION
.PP
TkRev is a Tcl/Tk-based graphical interface to the CVS, Subversion and Git configuration management systems. It displays the status of the files in the current working directory, and provides buttons and menus to execute configuration-management commands on the selected files. Limited RCS functionality is also present. Git functionality is new in version 9.
TkDiff is bundled in for browsing and merging your changes.
TkRev also aids in browsing the repository. For Subversion, the repository tree looks like an ordinary file tree. For CVS, the CVSROOT/modules file is read. TkRev extends CVS with a method to produce a browsable, "user friendly" listing of modules. This requires special comments in the CVSROOT/modules file. See the \fBCVS Modules File\fP section for more guidance.
.SH OPTIONS
.PP
TkRev accepts the following options

.TP
.B -dir \fIdirectory\fP
Start TkRev in the specified directory

.TP
.B -help
Print a usage message

.TP
.B -log \fIfile\fP
Invoke a log browser for the specified file

.TP
.B -blame \fIfile\fP
Invoke a blame (annotation) browser for the specified file

.TP
.B -root \fIcvsroot\fP
Set $CVSROOT to the specified repository.

.TP
.B -win  \fBworkdir\fP|\fBmodule\fP|\fBmerge\fP
Start by displaying the directory browser (the default), the module browser, or the directory-merge tool. -win and -log are mutually exclusive.
.PP
.SS Examples
Browse the modules located in CVSROOT /jaz/repository:

% tkrev -win module -root /jaz/repository

View the log of the file tstheap.c:

% tkrev -log tstheap.c
.SH Current Directory Browser
.PP
The working directory browser shows the files in your local working copy, or "sandbox."  It shows the status of the files at a glance and provides tools to help with most of the common version control operations you might do.

At the top of the browser you will find:

\fB*\fP The name of the current directory. You can change directories by typing in this field. Recently visited directories are saved in the picklist.

\fB*\fP The relative path of the current directory in the repository. If it is not contained in the repository you may import it using the menu or toolbar button.

\fB*\fP A Directory Tag name, if the directory is contained in the repository and it has been checked out against a particular branch or tag. In Subversion, the branch or tag is inferred from the URL based on the conventional trunk-branches-tags repository organization.

\fB*\fP The repository location of the current directory - CVSROOT if it's under CVS control, the URL of the Subversion repository if it's under Subversion control, or the origin if it's controlled by Git. If not a version-controlled directory, it may default to the value of the $CVSROOT environment variable.

The main part of the working directory browser is a list of the files in the current directory with an icon next to each showing its status. You select a file by clicking on its name or icon once with the left mouse button. Holding the Control key while clicking will add the file to the group of those already selected. You can select a contiguous group of files by holding the Shift key while clicking. You can also select a group of files by dragging the mouse with the middle or right button pressed to select an area. Selecting an item that's already selected de-selects that item. To unselect all files, click the left mouse button in an empty area of the file list.

\fB*\fP The Date column (can be hidden) shows the modification time of the file is shown. The format of the date column may be specified with cvscfg(dateformat). The default format was chosen because it sorts the same way alphabetically as chronologically.

If the directory belongs to a revision system, other columns are present.

\fB*\fP The revision column shows which revision of the file is checked out, and whether it's on the trunk or on a branch.

\fB*\fP The status column (can be hidden) shows the revision of the file spelled out in text. This information is mostly redundant to the icon in the file column.

\fB*\fP The Editor/Author/Locker column (can be hidden) varies according to revision system. In Subversion, the author of the most recent checkin is shown. In CVS, it shows a list of people editing the files if your site uses "cvs watch" and/or "cvs edit". Otherwise, it will be empty. In RCS, it shows who, if anyone, has the file locked.

The optional columns can be displayed or hidden using the Options menu.

You can move into a directory by double-clicking on it.

Double clicking on a file will load the file into a suitable editor so you can change it. A different editor can be used for different file types (see Configuration Files).
.PP
.SS File Status

When you are in a directory that is under CVS, Subversion, or Git control, the file status will be shown by an icon next to each file. Checking the "Status Column" option causes the status to be displayed in text in its own column. Some possible statuses are:

.TP
.B Up-to-date
The file is up to date with respect to the repository.

.TP
.B Locally Modified
The file has been modified in the current directory since being checked out of the repository.

.TP
.B Locally Added
The file has been added to the repository. This file will become permanent in the repository once a commit is made.

.TP
.B Locally Removed
You have removed the file with remove, and not yet committed your changes.

.TP
.B Needs Checkout
Someone else has committed a newer revision to the repository. The name is slightly misleading; you will ordinarily use update rather than checkout to get that newer revision.

.TP
.B Needs Patch
Like Needs Checkout, but the CVS server will send a patch rather than the entire file. Sending a patch or sending an entire file accomplishes the same thing.

.TP
.B Needs Merge
Someone else has committed a newer revision to the repository, and you have also made modifications to the file.

.TP
.B Unresolved Conflict
This is like Locally Modified, except that a previous update command gave a conflict. You need to resolve the conflict before checking in.

.TP
.B ?
The file is not contained in the repository. You may need to add the file to the repository by pressing the "Add" button.

.TP
.B [directory:CVS]
A directory which has been checked out from a CVS repository.

.TP
.B [directory:SVN]
A directory which has been checked out from a Subversion repository. In Subversion, directories are themselves versioned objects.

.TP
.B [directory:RCS]
A directory which contains an RCS sub-directory or some files with the ,v suffix, presumably containing some files that are under RCS revision control.

.TP
.B [directory:GIT]
A directory which has been cloned from a Git repository.

.TP
.B [directory]
A directory not controlled by one of the supported revision control systems
.PP
.SS File Filters

.TP
.B Clean
You can specify file matching patterns to instruct TkRev which files you wish to see. You can also specify patterns telling it which files to remove when you press the "Clean" button or select the \fBFile -> Cleanup\fP menu item.

.TP
.B Hide
"Hide" works exactly the way a .cvsignore file works. That is, it causes non-CVS files with the pattern to be ignored. It's meant for hiding .o files and such. Any file under CVS control will be listed anyway.

.TP
.B Show
"Show" is the inverse. It hides non-CVS files except for those with the pattern.

.SS Buttons

.TP
.B Module Browser:
The big button at the upper right opens the module browser opens a module browser window which will enable you to explore items in the repository even if they're not checked out. In CVS, this requires that there be entries in the CVSROOT/modules file. Browsing can be improved by using TkRev-specific comments in CVSROOT/modules.

.TP
.B Go Up:
The button to the left of the entry that shows the current directory. Press it and you go up one level.
.PP
There are a number of buttons at the bottom of the window. Pressing on one of these causes the following actions:

.TP
.B Delete:
Press this button to delete the selected files. The files will not be removed from the repository. To remove the files from the repository as well as delete them, press the "Remove" button instead.

.TP
.B Edit:
Press this button to load the selected files in to an appropriate editor.

.TP
.B View:
Press this button to view the selected files in a Tk text window. This can be a lot faster then Edit, in case your preferred editor is xemacs or something of that magnitude.

.TP
.B Refresh:
Press this button to re-read the current directory, in case the status of some files may have changed.

.TP
.B Status Check:
Shows, in a searchable text window, the status of all the files. By default, it is recursive and lists unknown (?) files. These can be changed in the Options menu.


.TP
.B Directory Branch Browser:
For merging the entire directory. In Subversion, it opens the Branch Browser for "."  In CVS, it chooses a "representative" file in the current directory and opens a graphical tool for directory merges.

.TP
.B Log (Branch) Browse:
This button will bring up the log browser window for each of the selected files in the window. See the \fBBranch Diagram Browser\fP section.

.TP
.B Annotate:
This displays a window in which the selected file is shown with the lines highlighted according to when and by whom they were last revised. In Subversion, it's also called "blame."

.TP
.B Diff:
This compares the selected files with the equivalent files in the repository. A separate program called "TkDiff" (also supplied with TkRev) is used to do this. For more information on TkDiff, see TkDiff's help menu.

.TP
.B Merge Conflict:
If a file's status says "Needs Merge", "Conflict", or is marked with a "C" in CVS Check, there was a difference which CVS needs help to reconcile. This button invokes TkDiff with the -conflict option, opening a merge window to help you merge the differences.

.TP
.B Check In:
This button commits your changes to the repository. This includes adding new files and removing deleted files. When you press this button, a dialog will appear asking you for the version number of the files you want to commit, and a comment. You need only enter a version number if you want to bring the files in the repository up to the next major version number. For example, if a file is version 1.10, and you do not enter a version number, it will be checked in as version 1.11. If you enter the version number 3, then it will be checked in as version 3.0 instead. It is usually better to use symbolic tags for that purpose.  If you use rcsinfo to supply a template for the comment, you must use an external editor. Set cvscfg(use_cvseditor) to do this. For checking in to RCS, an externel editor is always used.

.TP
.B Update:
This updates your sandbox directory with any changes committed to the repository by other developers.

.TP
.B Update with Options:
Allows you to update from a different branch, with a tag, with empty directories, and so on.

.TP
.B Add Files:
Press this button when you want to add new files to the repository. You must create the file before adding it to the repository. To add some files, select them and press the Add Files button. The files that you have added to the repository will be committed next time you press the Check In button. It is not recursive. Use the menu \fBCVS -> Add Recursively\fP for that.

.TP
.B Remove Files:
This button will remove files. To remove files, select them and press the Remove button. The files will disappear from the directory, and will be removed from the repository next time you press the Check In button. It is not recursive. Use the menu \fBCVS -> Remove Recursively\fP for that.

.TP
.B Tag:
This button will tag the selected files. In CVS, the \fB-F (force)\fP option will move the tag if it already exists on the file.

.TP
.B Branch Tag:
This button will tag the selected files, creating a branch. In CVS, the \fB-F (force)\fP option will move the tag if it already exists on the file.

.TP
.B Lock (CVS and RCS):
Lock an RCS file for editing. If cvscfg(cvslock) is set, lock a CVS file. Use of locking is philosophically discouraged in CVS since it's against the "concurrent" part of Concurrent Versioning System, but locking policy is nevertheless used at some sites. One size doesn't fit all.

.TP
.B Unlock (CVS and RCS):
Unlock an RCS file. If cvscfg(cvslock) is set, unlock a CVS file.

.TP
.B Set Edit Flag (CVS):
This button sets the edit flag on the selected files, enabling other developers to see that you are currently editing those files (See "cvs edit" in the CVS documentation).

.TP
.B Reset Edit Flag (CVS):
This button resets the edit flag on the selected files, enabling other developers to see that you are no longer editing those files (See "cvs edit" in the CVS documentation). As the current version of cvs waits on a prompt for "cvs unedit" if changes have been made to the file in question (to ask if you want to revert the changes to the current revision), the current action of tkrev is to abort the unedit (by piping in nothing to stdin). Therefore, to lose the changes and revert to the current revision, it is necessary to delete the file and do an update (this will also clear the edit flag). To keep the changes, make a copy of the file, delete the original, update, and then move the saved copy back to the original filename.

.TP
.B Close:
Press this button to close the Working Directory Browser. If no other windows are open, TkRev exits.
.SH Repository Browser
.PP
Operations that are performed on the repository instead of in a checked-out working directory are done with the Module Browser. The most common of these operations is checking out or exporting from the repository. The Module Browser can be started from the command line (tkrev -win module) or started from the main window by pressing the big button.

Subversion repositories can be browsed like a file tree, and that is what you will see in the Module Browser. CVS repositories aren't directly browsable, but if the CVSROOT/modules file is maintained appropriately, TkRev can display the modules and infer tree structures if they are present. See the \fBCVS Modules File\fP section.

Using the module browser window, you can select a module to check out. When you check out a module, a new directory is created in the current working directory with the same name as the module.

.SS Tagging and Branching (cvs rtag)

You can tag particular versions of a module or file in the repository, with plain or branch tags, without having the module checked out.

.SS Exporting

Once a software release has been tagged, you can use a special type of checkout called an export. This allows you to cleanly check out files from the repository,  without all of the administrivia that CVS needs to have while working on the files. It is useful for delivery of a software release to a customer.

.SS Importing

TkRev contains a special dialog to allow users to import new files into the repository. In CVS, new modules can be assigned places within the repository, as well as descriptive names (so that other people know what they are for).


CHANGELOG
Release 8.2.2 May, 2010

- Make the propget svn:mergeinfo branch diagram behave more like our merge
  tags, i.e. show only the first merge point instead of all revisions
  containing the merge.  This should improve performance relative to version
  8.2, too.

- Improve performance in large Subversion directories

- Fix CDE font problem

- Show date of directories, too.  Fix sort-by-revision in SVN directory.

- Roll back svn update --accept postpone because it doesn't work with older
  clients

- Clean up finished namespaces for some exec viewers


Release 8.2.1 January, 2010

- History of commit log messages is kept, so you can copy-and-paste previous
  messages that you've used during the current session.

- When a directory is refreshed in the Working Directory Browser, the scroll
  position is restored.

- Locking and unlocking for SVN, and some additional SVN-specific right-click
  popups (Matthias Vorwerk)

- Icons for symbolic links in a SVN directory (Matthias Vorwerk)

- SVN output parsing change (xml-regexp-based) (Matthias Vorwerk)

- Subversion Branch diagram can find branchpoint for a file that wasn't
  revised at that point (Steve Schwarm)

- CVS branch browser no longer fooled by a log comment containing a dashed line
  of exactly the same length as the log's normal delimiter, although you may
  not see the rest of the comment

- tksvn2bcompare.pl added to contrib directory.  It enables the use of
  "bcompare" in place of tkdiff.  (Adam McLaurin)

- Add a note to the FAQ about running the X11 version of TkCVS on the Mac.

- Fix mixed-up tooltips in the File Browser opened from the Module Browser.

- Fix Reports->Status recursive/local switch on cvscfg(recurse)

- Button in the CVS "Update with Options" dialog to make it easier to update
  a sticky-tagged file to the current directory tag (Jacques Klein)

- Change tkcvs.tcl so it can be made into a starkit more easily


Release 8.2 November, 2008

- Merge arrows are drawn in the Branch Browser for merges tracked by
  Subversion 1.5's mergeinfo property and CVSNT's mergepoint feature.

- The branch diagram can be searched to find a version, date, tag, or author

- Log browser always produces a verbose log of revisions on the selected branch
  instead of obeying the Directory Browser's "Log Detail" setting

- If your SVN repository has a structure similar to trunk, branches, and tags
  but with different names, you can tell TkCVS about it by setting variables in
  tkcvs_def.tcl:
    set cvscfg(svn_trunkdir) "elephants"
    set cvscfg(svn_branchdir) "dogs"
    set cvscfg(svn_tagdir) "ducklings"

- Fix a bash-ism in contrib/cvsdiff

- Changed the trace levels so that "F" lets you get the CVS/SVN stdout without
  the whole debug output


Release 8.1   November, 2007

- Rework the merge functionality.  There's only one dialog for tagging, which
  you OK when you're ready to commit the merges.

- Use panedwindow for the Workdir Browser. It has advantages and disadvantages,
  but it will have more advantages when we can migrate to tk8.5.  Change the
  highlighting so it goes across all columns, and enable selection from all
  columns.

- Add a menu item to do "svn resolved"

- Fix [ 1824733 ] CVS menu in SVN work area for changed file

- Fix invocation of tkdiff when one SVN revision is selected in the branch
  browser, diffing it against the current file like the cvs behavior

- Add options to use -l and not use -P in cvs update-with-options.

- Improve visibility of searched item in annotation text.

- Fix for when an e-mail address appears in svn status in https protocol

- Remove white boxes around Aqua pill-style buttons


Release 8.0.4   May, 2007

- The Branch Browser detects lack of a trunk directory, warns that
  it can't do much without that structure, and continues without it.

- Fix [ 1483057 ] Empty Branch Diagram for deleted files (TkSVN)
  If a file had been removed from the trunk and was diagrammed from a
  branch, some or all of the diagram could be missing.

- Fix [ 1673519 ] tcl error with SVN->Browse the Log Diagram

- Fix [ 1581111 ] svn url trouble with French localization

- Ask for confirmation before reverting files

- Display TkCVS version in window title of workdir and module browsers.

- Don't fail if CVS gives a date format that tcl can't handle

- Added a button for a text history log of the file to the branch browser.

- Added a command line option -annotate or -blame to open the annotation
  browser from the command line

- The "New Directory" button is back

- cvscfg(svn_branch_filter) and cvscfg(svn_branch_max_count) to filter
  which branches to draw in the log browser

- Delete the exec namespaces after the execs are finished.  This should
  cause better memory usage behavior.

- Add a contrib directory containing a wrapper for gvimdiff to replace
  tkdiff, and a program to compare the contents of directories.


Release 8.0.3   March, 2006

- Automatic tagging of merges works for SVN the same as CVS

- Working directory browser observes svn_ignore

- Clean up some filenames-with spaces issues

- TkDiff 4.1.3


Release 8.0.2   January, 2006

- Fix error in Branch Diagram when searching for merge tags

- cvscfg(mergetrunkname) option to replace the literal "trunk" in the code
  with an arbitrary string

- Branch Browser in SVN will diff a single selection in the tree with the
  file in the current directory

- Fix strange "SVN Path" in top entry of Branch Browser (only cosmetic)

- Lengthen maximum length of error message to trigger an error popup in
  exec.  That lets a cvs log failure due to a permission problem tell us
  what went wrong.


Release 8.0.1   January, 2006

- Fix a couple of undefined variables

- Add log button to workdir browser and change the cvs_log function to
  eliminate post-processing, using syntax highlighting instead


Release 8.0     December, 2005

- The Annotation browser optionally shows line numbers.

- Multiple branch-browser fixes for Subversion: 

trunk/tkrev/tkrev.tcl
#!/bin/sh
#-*-tcl-*-
# the next line restarts using wish \
    exec wish "$0" -- ${1+"$@"}

#
# TkRev Main program -- A Tk interface to CVS.
#
# Uses a structured modules file -- see the manpage for more details.
#
# Author:  Del (del@babel.dialix.oz.au)
#

# If we can't get this far (maybe because X display connection refused)
# quit now.  If we get further, the error message is very misleading.
if {[info exists starkit::topdir]} {
  package require Tk
}

if {! [info exists tk_version] } {
  puts "Initialization failed"
  exit 1
}

if {$tk_version < 8.5} {
  puts "TkRev requires Tcl/Tk 8.5 or better!"
  exit 1
}

if {[info exists starkit::topdir]} {
  set TclRoot [file join $starkit::topdir lib]
  set ScriptBin $starkit::topdir
} else {
  if {[info exists TclRoot]} {
    # Perhaps we are being sourced recursively.
    # That would be bad.
    return
  }
  set Script [info script]
  set ScriptTail [file tail $Script]
  #puts "Tail $ScriptTail"
  if {[file type $Script] == "link"} {
    #puts "$Script is a link"
    set ScriptBin [file join [file dirname $Script] [file readlink $Script]]
  } else {
    set ScriptBin $Script
  }
  #puts "  ScriptBin $ScriptBin"
  set TclRoot [file join [file dirname $ScriptBin]]
  #puts "TclRoot $TclRoot"
  if {$TclRoot == "."} {
    set TclRoot [pwd]
  }
  #puts "TclRoot $TclRoot"
  set TclRoot [file join [file dirname $TclRoot] "lib"]
  #puts "TclRoot $TclRoot"
  
  # allow runtime replacement
  if {[info exists env(TCLROOT)]} {
    set TclRoot $env(TCLROOT)
  }
  #puts "TclRoot $TclRoot"
}

set TclExe [info nameofexecutable]
if {$tcl_platform(platform) == "windows"} {
  set TclExe [file attributes $TclExe -shortname]
}


set TCDIR [file join $TclRoot tkrev]
set cvscfg(bitmapdir) [file join $TclRoot tkrev bitmaps]
#puts "TCDIR $TCDIR"
#puts "BITMAPDIR $cvscfg(bitmapdir)"

set cvscfg(version) "9.4"

if {! [info exists cvscfg(editorargs)]} {
  set cvscfg(editorargs) {}
}
set auto_path [linsert $auto_path 0 $TCDIR]
set cvscfg(allfiles) false
if {! [info exists cvscfg(startwindow)]} {
  set cvscfg(startwindow) "workdir"
}
set cvscfg(auto_tag) false
set cvscfg(econtrol) false
set cvscfg(use_cvseditor) false

set maxdirs 15
set dirlist {}
set totaldirs 0

# Orient ourselves
if { [info exists env(HOME)] } {
  set cvscfg(home) $env(HOME)
} else {
  set cvscfg(home) "~"
}
if { [info exists env(USER)] } {
  set cvscfg(user) $env(USER)
} elseif { [info exists env(USERNAME)] } {
  # Windows
  set cvscfg(user) $env(USERNAME)
} else {
  set cvscfg(user) ""
}

# Read in defaults
if {[file exists [file join $TCDIR tkrev_def.tcl]]} {
  source [file join $TCDIR tkrev_def.tcl]
}

set optfile [file join $cvscfg(home) .tkrev]
set old_optfile [file join $cvscfg(home) .tkcvs]
if {[file exists $old_optfile] && ![file exists $optfile]} {
  file copy $old_optfile $optfile
}
if {[file exists $optfile]} {
  catch {source $optfile}
}
set pickfile [file join $cvscfg(home) .tkrev-picklists]
if {[file exists $pickfile]} {
  picklist_load
}
if {! [info exist cvsglb(directory)]} {
  set cvsglb(directory) [pwd]
}
if {[info exists cvsglb(cvsroot)]} {
  set cvscfg(cvsroot) [lindex $cvsglb(cvsroot) 0]
}

# Set some defaults
set cvsglb(commit_comment) ""
set cvsglb(cvs_version) ""

if {$cvscfg(use_cvseditor) && ![info exists cvscfg(terminal)]} {
  cvserror "cvscfg(terminal) is required if cvscfg(use_cvseditor) is set"
}

# Hilight colors.  Get the colorful ones.
entry .testent
wm withdraw .
set cvsglb(textbg) white
set cvsglb(textfg) black
set cvsglb(hlbg) [lindex [.testent configure -selectbackground] 4]
set cvsglb(hlfg) [lindex [.testent configure -selectforeground] 4]
if {$cvsglb(hlfg) eq {} } {
  # This happens on the Mac
  set cvsglb(hlfg) [lindex [.testent configure -foreground] 4]
}
set cvscfg(listboxfont) [lindex [.testent configure -font] 4]
#puts [font actual $cvscfg(listboxfont) -displayof .testent]
#puts [font metrics $cvscfg(listboxfont) -displayof .testent]
destroy .testent


set WSYS [tk windowingsystem]
#puts "Windowing sytem is $WSYS"
set theme_system "unknown"

if {$WSYS eq "x11"} {
  # If X11, see if we can sense our environment somehow
  label .testlbl -text "LABEL"
  if [get_cde_params] {
    set theme_system "CDE"
    # Find out what the default gui font is
    if { ! [info exists cvscfg(guifont)] } {
      # Find out what the tk default is
      set cvscfg(guifont) [lindex [.testlbl configure -font] 4]
    }
    # Put the Help menu back on the right
    tk::classic::restore menu
    # canvbg is for the treeviews, not the branch browser canvas
    set cvsglb(canvbg) $cvsglb(shadow)
  } elseif [get_gtk_params] {
    set theme_system "GTK"
    if { ! [info exists cvscfg(guifont)] } {
      set cvscfg(guifont) [lindex [.testlbl configure -font] 4]
      font configure TkDefaultFont -size 9
      set cvscfg(guifont) TkDefaultFont
      option add *Menu.font $cvscfg(guifont)
      option add *Label.font $cvscfg(guifont)
      option add *Button.font $cvscfg(guifont)
    }
    # in KDE or Gnome or some such.  It rather rudely sets all the Tk
    # backgrounds the same which I don't like, so I'm going to use the same
    # trick I use for CDE to give the canvases a little shading.
    set cvsglb(bg) [lindex [.testlbl cget -background] 0]
    set cvsglb(fg) [lindex [.testlbl cget -foreground] 0]
    # canvbg is for the treeviews, not the branch browser canvas
    set cvsglb(canvbg) [rgb_shadow $cvsglb(bg)]
  } else {
    set bg  [lindex [.testlbl cget -background] 0]
    set fg  [lindex [.testlbl cget -foreground] 0]
    set hlbg "#4a6984"
    set hlfg "#ffffff"
    set textbg "#ffffff"
    set textfg "#000000"
    
    shades $bg
    
    set cvsglb(bg) $bg
    set cvsglb(fg) $fg
    set cvsglb(textbg) $textbg
    set cvsglb(textfg) $textfg
    set cvsglb(hlbg) $hlbg
    set cvsglb(hlfg) $hlfg
    
    option add *Canvas.Background $cvsglb(shadow)
    option add *Canvas.Foreground black
    option add *Entry.Background $textbg
    option add *Entry.Foreground $textfg
    option add *Entry.selectBackground $hlbg
    option add *Entry.selectForeground $hlfg
    option add *Entry.readonlyBackground $cvsglb(lighter)
    option add *Listbox.background $textbg
    option add *Listbox.selectBackground $hlbg
    option add *Listbox.selectForeground $hlfg
    option add *Text.Background $textbg
    option add *Text.Foreground $textfg
    option add *Text.selectBackground $hlbg
    option add *Text.selectForeground $hlfg
    option add *Button.activeForeground $fg
    option add *Menu.activeForeground $fg
    option add *Checkbutton.Background $bg
    
    # checkbuttons and radiobuttons
    option add *Menu.selectColor $fg
    option add *Checkbutton.selectColor "#ffffff"
    option add *Radiobutton.selectColor "#ffffff"
    
    if { ! [info exists cvscfg(guifont)] } {
      set cvscfg(guifont) [lindex [.testlbl configure -font] 4]
      # This makes it look more classic
      font configure TkHeadingFont -size 9
      set cvscfg(guifont) TkHeadingFont
      option add *Menu.font $cvscfg(guifont)
      option add *Label.font $cvscfg(guifont)
      option add *Button.font $cvscfg(guifont)
    }
  }
  destroy .testlbl
  
  if {! [info exists cvscfg(dialogfont)]} {
    set cvscfg(dialogfont) $cvscfg(guifont)
  }
  
  if {$theme_system == "CDE"} {
    # This makes it consistent with the rest of the CDE interface
    option add *Menu.font $cvscfg(guifont)
    option add *Label.font $cvscfg(guifont)
    option add *Button.font $cvscfg(guifont)
  }
  #puts " Theme system: $theme_system"
} else {
  # Find out what the default gui font is
  label .testlbl -text "LABEL"
  # Find out what the tk default is
  set cvscfg(guifont) [lindex [.testlbl configure -font] 4]
  set cvscfg(dialogfont) $cvscfg(guifont)
  
  set cvsglb(bg) [lindex [.testlbl cget -background] 0]
  set cvsglb(fg) [lindex [.testlbl cget -foreground] 0]
  # canvbg is for the treeviews, not the branch browser canvas
  set cvsglb(canvbg) [rgb_shadow $cvsglb(bg)]
  destroy .testlbl
  if {$WSYS eq "aqua"} {
    # Keep everything from being blinding white
    set arbitrarybg "#dddddd"
    set lighterbg "#eeeeee"
    option add *Frame.background $arbitrarybg userDefault
    option add *Label.background $arbitrarybg userDefault
    option add *Entry.highlightBackground $arbitrarybg userDefault
    option add *Entry.readonlyBackground $lighterbg userDefault
    option add *Canvas.highlightBackground #fefefe userDefault
    option add *Message.Background $arbitrarybg userDefault
    option add *Checkbutton.Background $arbitrarybg userDefault
    option add *Radiobutton.Background $arbitrarybg userDefault
    # button highlightbackground has to be the same as background
    # or else there are little white boxes around the button "pill"
    option add *Button.highlightBackground $arbitrarybg userDefault
    # canvbg is for the treeviews, not the branch browser canvas
    set cvsglb(canvbg) $lighterbg
  }
}

# Suppress tearoffs in menubars
option add *tearOff 0

option add *ToolTip.background  "LightGoldenrod1" userDefault
option add *ToolTip.foreground  "black" userDefault

# This makes tk_messageBox use our font.  The default tends to be terrible
# no matter what platform
option add *Dialog.msg.font $cvscfg(dialogfont) userDefault
# Sometimes we'll want to override this but let's set a default
option add *Message.font $cvscfg(dialogfont) userDefault

if {$WSYS eq "x11"} {
  ttk::style configure TCombobox -arrowsize 16
  # Header padding has no effect on aqua, but it works on X11
  ttk::style configure Treeview.Heading -padding {4 0}
}
# And, background seems to have no effect on X11
ttk::style configure Treeview -font $cvscfg(listboxfont) -background $cvsglb(canvbg) \
    -fieldbackground $cvsglb(canvbg)
ttk::style configure Treeview.Heading -font $cvscfg(listboxfont) -background $cvsglb(bg)
ttk::style configure Treeview.Cell -padding {2 0}

# Initialize logging (classes are C,F,T,D)
if { ! [info exists cvscfg(log_classes)] } {
  set cvscfg(log_classes) "C"
}
foreach class [split $cvscfg(log_classes) {}] {
  set logclass($class) $class
}
if { ! [info exists cvscfg(logging)] } {
  set cvscfg(logging) false
}
load_all_images
if {$cvscfg(logging)} {
  gen_log:init
}
# Detect whether we're in a revision-controlled directory
lassign [cvsroot_check [pwd]] incvs insvn inrcs ingit

#
# Command line options
#
set usage "Usage:"
append usage "\n tkrev \[-root <cvsroot>\] \[-win workdir|module|merge\]"
append usage "\n tkrev \[-log|blame <file>\]"
append usage "\n tkrev <file> - same as tkrev -log <file>"
for {set i 0} {$i < [llength $argv]} {incr i} {
  set arg [lindex $argv $i]
  set val [lindex $argv [expr {$i+1}]]
  switch -regexp -- $arg {
    {^--*d.*} {
      # -ddir: Starting directory
      set dir $val; incr i
      cd $val
    }
    {^--*r.*} {
      # -root: CVS root
      set cvscfg(cvsroot) $val; incr i
    }
    {^--*w.*} {
      # workdir|module|merge: window to start with. workdir is default.
      set cvscfg(startwindow) $val; incr i
    }
    {^--*l.*} {
      # -log <filename>: Browse the log of specified file
      set cvscfg(startwindow) log
      set lcfile $val; incr i
    }
    {^--*[ab].*} {
      # annotate|blame: Browse colorcoded history of specified file
      set cvscfg(startwindow) blame
      set lcfile $val; incr i
    }
    {^-psn_.*} {
      # Ignore the Carbon Process Serial Number
      incr i
    }
    {^--*h.*} {
      puts $usage
      exit 0
    }
    {^\.*\w*} {
      # If a filename is provided as an argument, assume -log
      # except if it's a directory and it's CVS, which doesn't
      # version directories
      if {($insvn || $ingit)} {
        set cvscfg(startwindow) log
        set lcfile $arg; incr i
      } else {
        if {[file isdirectory $arg]} {
          set dir $arg
          cd $arg
        } else {
          set cvscfg(startwindow) log
          set lcfile $arg; incr i
        }
      }
    }
    default {
      puts $usage
      exit 1
    }
  }
}

if {[info exists lcfile]} {
  set d [file dirname $lcfile]
  set f [file tail $lcfile]
  set lcfile $f
  cd $d
}

if {![info exists cvscfg(ignore_file_filter)]} {
  set cvscfg(ignore_file_filter) ""
}
if {[info exists cvscfg(file_filter)]} {
  unset cvscfg(file_filter)
}
if {![info exists cvscfg(show_file_filter)]} {
  set cvscfg(show_file_filter) "*"
}


set cvsglb(root) ""
set cvsglb(vcs) ""

# Create a window
# Start with Module Browser
if {[string match {mod*} $cvscfg(startwindow)]} {
  wm withdraw .
  # If we're in a version-controlled directory, open that repository
  if {$insvn} {
    set cvsglb(root) $cvscfg(svnroot)
    set cvsglb(vcs) svn
  } elseif {$incvs} {
    set cvsglb(root) $cvscfg(cvsroot)
    set cvsglb(vcs) cvs
  } elseif {$ingit} {
    set cvsglb(root) $cvscfg(url)
    set cvsglb(vcs) git
  } else {
    # We'll respect CVSROOT environment variable if it's set
    if {[info exists env(CVSROOT)]} {
      set cvsglb(root) $env(CVSROOT)
      set cvscfg(cvsroot) $env(CVSROOT)
      set cvsglb(vcs) cvs
    }
  }
  # Othewise we set it to the most recent saved in picklist
  # which we've saved in cvscfg(cvsroot)
  if {$cvsglb(root) == ""} {
    set cvsglb(root) $cvscfg(cvsroot)
  }
  modbrowse_run
  # Start with Branch Browser
} elseif {$cvscfg(startwindow) == "log"} {
  if {! [file exists $lcfile]} {
    puts "ERROR: $lcfile doesn't exist!"
    exit 1
  }
  wm withdraw .
  if {$incvs} {
    cvs_branches [list $lcfile]
  } elseif {$inrcs} {
    set cwd [pwd]
    set module_dir ""
    rcs_branches [list $lcfile]
  } elseif {$insvn} {
    svn_branches [list $lcfile]
  } elseif {$ingit} {
    git_branches [list $lcfile]
  } else {
    puts "File doesn't seem to be in CVS, SVN, RCS, or GIT"
  }
  # Start with Annotation Browser
} elseif {$cvscfg(startwindow) == "blame"} {
  if {! [file exists $lcfile]} {
    puts "ERROR: $lcfile doesn't exist!"
    exit 1
  }
  wm withdraw .
  if {$incvs} {
    cvs_annotate $current_tagname [list $lcfile]
  } elseif {$insvn} {
    svn_annotate rBASE [list $lcfile]
  } elseif {$ingit} {
    read_git_dir .
    git_annotate $current_tagname [list $lcfile]
  } else {
    puts "File doesn't seem to be in CVS, SVN, or GIT"
  }
  # Start with Directory Merge
} elseif {[string match {mer*} $cvscfg(startwindow)]} {
  wm withdraw .
  if {$incvs} {
    cvs_joincanvas
  } elseif {$insvn} {
    svn_directory_merge
  } else {
    puts "Directory doesn't seem to be in CVS or SVN"
  }
  # The usual way, with the Workdir Browser
} else {
  setup_dir
}

